src/with_position.rs

```diff
@@ -81,6 +81,34 @@ impl<I: Iterator> Iterator for WithPosition<I> {
     fn size_hint(&self) -> (usize, Option<usize>) {
         self.peekable.size_hint()
     }
+
+    fn fold<B, F>(mut self, mut init: B, mut f: F) -> B
+    where
+        F: FnMut(B, Self::Item) -> B,
+    {
+        if !self.handled_first {
+            // Haven't seen the first item yet, and there might be one to give.
+            self.handled_first = true;
+            if let Some(first) = self.peekable.next() {
+                // Peek to see if this is also the last item,
+                // in which case tag it as `Only`.
+                match self.peekable.peek() {
+                    Some(_) => init = f(init, (Position::First, first)),
+                    None => return f(init, (Position::Only, first)),
+                }
+            }
+        }
+        if let Some(mut head) = self.peekable.next() {
```

Comment:
I'm somehow bothered because we call `peekable.next` here, even though we called `peekable.peek` some lines before. I'm not sure if we can easily avoid this (without convoluting control flow too much)...

Maybe something like this (only a draft - didn't test it):
````
    if let Some(head) = if !handled_first {
        if let Some(first) = self.peekable.next() {
            match self.peekable.next() {
                Some(head) => {
                    init = {
                        f(init, (Position::First, first));
                        Some(head)
                    }
                }
                None => return f(init, (Position::Only, first)),
            }
        }
    } else {
        self.peekable.next()
    } {
        // Have seen the first item, and there's something left.
        init = self.peekable.fold(init, |acc, mut item| {
            std::mem::swap(&mut head, &mut item);
            f(acc, (Position::Middle, item))
        });
        // The "head" is now the last item.
        init = f(init, (Position::Last, head));
    }
````